cloud2sql/parquet.py

```diff
@@ -0,0 +1,158 @@
+            from_table = get_table_name(from_kind, with_tmp_prefix=False)
+            to_table = get_table_name(to_kind, with_tmp_prefix=False)
+            link_table = get_link_table_name(from_kind, to_kind, with_tmp_prefix=False)
+            if link_table not in self.schemas and from_table in self.schemas and to_table in self.schemas:
+                schema = pa.schema(
+                    [
+                        pa.field("from_id", pa.string()),
+                        pa.field("to_id", pa.string()),
+                    ]
+                )
+                self.schemas[link_table] = schema
+
+        def link_table_schema_from_successors(kind: Kind) -> None:
+            _, successors = kind_properties(kind, self.model)
+            # create link table for all linked entities
+            for successor in successors:
+                link_table_schema(kind.fqn, successor)
+
+        # step 1: create tables for all kinds
+        for kind in self.table_kinds:
+            table_schema(kind)
+        # step 2: create link tables for all kinds
+        for kind in self.table_kinds:
+            link_table_schema_from_successors(kind)
+        # step 3: create link tables for all seen edges
+        for from_kind, to_kind in edges:
+            link_table_schema(from_kind, to_kind)
+
+        return None
+
+
+class WriteResult(NamedTuple):
+    table_name: str
+
+
+@dataclass
+class ParquetBatch:
+    rows: List[Dict[str, Any]]
+    schema: pa.Schema
+    writer: pq.ParquetWriter
+
+
+class ParquetWriter:
+    def __init__(
+        self,
+        model: ParquetModel,
+        result_directory: Path,
+        rows_per_batch: int,
+    ):
+        self.model = model
+        self.kind_by_id: Dict[str, str] = {}
+        self.batches: Dict[str, ParquetBatch] = {}
+        self.rows_per_batch = rows_per_batch
+        self.result_directory = result_directory
+
+    def insert_value(self, table_name: str, values: Any) -> Optional[WriteResult]:
+        if self.model.schemas.get(table_name):
+
+            def ensure_path(path: Path) -> Path:
+                path.mkdir(parents=True, exist_ok=True)
+                return path
+
+            batch = self.batches.get(
+                table_name,
+                ParquetBatch(
+                    [],
+                    self.model.schemas[table_name],
+                    pq.ParquetWriter(
+                        Path(ensure_path(self.result_directory), f"{table_name}.parquet"),
+                        self.model.schemas[table_name],
+                    ),
+                ),
+            )
+
+            batch.rows.append(values)
+            self.batches[table_name] = batch
+            return WriteResult(table_name)
+        return None
+
+    def write_batch_bundle(self, batch: ParquetBatch) -> None:
+        pa_table = pa.Table.from_pylist(batch.rows, batch.schema)
+        batch.writer.write_table(pa_table)
+        batch.rows = []
```

Comment:
I guess there is no concurrency here? Consider assigning the rows to a local variable and emptying the batch as first action.